torch_xla/experimental/fori_loop.py

```diff
@@ -10,80 +10,139 @@
 from torch._ops import HigherOrderOperator
 import torch._higher_order_ops.while_loop
 from torch._higher_order_ops.while_loop import while_loop_op
+from torch._higher_order_ops.while_loop import while_loop as torch_while_loop
+from torch._higher_order_ops.utils import _has_potential_branch_input_mutation
 
 
-def fori_loop(lower, upper, user_body_func, *init_val):
+def fori_loop(upper, lower, body_fun, *input_value):
 
   device = xm.xla_device()
+  if (upper < lower):
+    print("ERROR: upper should be a larger number than lower")
+  iteri = upper - lower
 
-  def cond_fn(upper, lower, *init_val):
-    return lower[0] < upper[0]
+  def cond_fn(iteri, *input_value):
+    return iteri > 0
 
-  def body_fn(upper, lower, *init_val):
-    one_value_i = torch.ones(1, dtype=torch.int32, device=device)
-    res_list = list(user_body_func(*init_val))
-    res_list.insert(0, lower)
-    res_list.insert(0, torch.sub(upper, one_value_i))
-    return res_list
+  def new_body_fn(iteri, *input_value):
+    return iteri - 1, body_fun(*input_value)
+
+  inputs = (iteri,) + input_value
+  res = _xla_while_loop_wrapper(
+      cond_fn, new_body_fn, inputs, (), fake_tensor=True)
 
-  res = while_loop(cond_fn, body_fn, (lower, upper, *init_val))
   return res
 
 
 @while_loop_op.py_impl(DispatchKey.XLA)
-def while_loop(cond_fn, body_fn, *carried_inputs, additional_inputs=None):
-  # TODO(@manfei): PyTorch require carried_inputs to be list/tuple, PyTorch/XLA _xla_while_loop only accept *operands, *operands would tuple items again: (a, '')
-  # cond_fn&body_fn: callable
-  # carried_inputs: (Tuple of possibly nested dict/list/tuple of tensors)
+def while_loop(cond_fn, body_fn, carried_inputs, additional_inputs=None):
   if additional_inputs is None:
     additional_inputs = tuple()
-  return _xla_while_loop(
-      cond_fn, body_fn, *carried_inputs, additional_inputs=additional_inputs)
+  return _xla_while_loop_wrapper(cond_fn, body_fn, carried_inputs,
+                                 additional_inputs)
+
+
+def _xla_while_loop_wrapper(cond_fn,
+                            body_fn,
+                            carried_inputs,
+                            additional_inputs=None,
+                            fake_tensor=False):
+
+  def new_body_fn(*carried_inputs):
+    res = list(body_fn(*carried_inputs))
+    if additional_inputs:
+      res = [
+          res[0],
+      ] + list(additional_inputs) + res[1:]
```

Comment:
what does these lines do?

Reply:
`this line` is to create return value based on post-order rule: `iterator` + `weight/bias` + `inputs`

### detail:
the original gain body's xlacomputation like:
```python
in: {(iterator, weight1, bias1, weight2, bias2, inputs)}

res = body(inputs)

return: {(iterator, res)}
```
(here example `res` and `inputs` has the same shape+type)

due to `xla::While` has requirement:`inputs/return should be the same shape+type`,

so `this line` is to create return value based on post-order rule: `iterator` + `weight/bias` + `inputs`:
- iterator: [res[0], ]
- weight/bias: list(additional_inputs)
- inputs: res[1:]

after this line, the new xlacomputation would like:
```python
in: {(iterator, weight1, bias1, weight2, bias2, inputs)}

res = body(inputs)

return: {(iterator, weight1, bias1, weight2, bias2, res)}
```